fuzzers/main.py

```diff
@@ -0,0 +1,642 @@
+                f"```\n"
+                f"\n"
+                f"Please edit the topic name to add more meaningful information once investigated. Keeping the artifact hash in it can help if the same artifact gets detected as crashing another branch.\n"
+            ),
+        })
+        last_log_lines = ""
+        for l in log_lines[::-1]:
+            if l.startswith("```"):
+                # Censor the end of a spoiler block, not great but this is for human consumption
+                # anyway
+                l = " " + l
+            if len(last_log_lines) + len(l) > 9000:
+                # Zulip limit is 10k, let's keep some safety buffer here
+                break
+            last_log_lines = l + last_log_lines
+        client.send_message({
+            "type": "stream",
+            "to": "nearinc/fuzzer/private",
+            "topic": f"{branch}: artifact {artifact}",
+            "content": f"```spoiler Last few log lines\n{last_log_lines}\n```",
+        })
+
+    def signal(self, signal: int) -> None:
+        """
+        Signal the fuzzer process (with all its process group as a fuzzer can spawn sub-fuzzers)
+        """
+
+        print(f"Sending signal {signal} to fuzzer {self.proc.pid}")
+        os.killpg(os.getpgid(self.proc.pid), signal)
+
+def random_weighted(array, num: int):
+    return random.choices(array, [x['weight'] for x in array], k = num)
+
+def pause_exit_spot(pause_evt: threading.Event, resume_evt: threading.Event, exit_evt: threading.Event) -> None:
+    """
+    Poll the events, returning True if `exit_evt` was set and pausing as requested by `pause_evt`
+    and `resume_evt`
+    """
+
+    resume_evt.clear()
+    if pause_evt.is_set():
+        while not resume_evt.wait(timeout=1):
+            if exit_evt.is_set():
+                return True
+        pause_evt.clear()
+    return exit_evt.is_set()
+
+def kill_fuzzers(bucket: gcs.Bucket, fuzzers: typing.Iterable[FuzzProcess]):
+    for f in fuzzers:
+        try:
+            f.signal(signal.SIGTERM)
+        except ProcessLookupError:
+            print(f"Failed looking up process {f.proc.pid}")
+    time.sleep(5)
+    for f in fuzzers:
+        try:
+            f.signal(signal.SIGKILL)
+        except ProcessLookupError:
+            print(f"Failed looking up process {f.proc.pid}")
+    for f in fuzzers:
+        bucket.blob(f"logs/{f.log_path}").upload_from_filename(LOGS_DIR / f.log_path)
+
+def run_fuzzers(gcs_client: gcs.Client, pause_evt: threading.Event, resume_evt: threading.Event, exit_evt: threading.Event):
+    """
+    Run all the fuzzers until `exit_evt` gets triggered, pausing and resuming
+    them based on `pause_evt` and `resume_evt`.
+    """
+
+    bucket = gcs_client.bucket(GCS_BUCKET)
+
+    repo = Repository(REPO_DIR, REPO_URL)
+    repo.clone_if_need_be()
+
+    corpus = Corpus(CORPUS_DIR, bucket)
+
+    while True:
+        if pause_exit_spot(pause_evt, resume_evt, exit_evt):
+            return
```

Comment:
In other words, `while not pause_exit_spot(pause_evt, resume_evt, exit_evt):` ;)

Reply:
That's possible too, here I wanted to semantically show that this is an infinite loop that can just be broken out by some other mechanism at designated points. Would you rather I reword the code?